2.6.0

KillerOfPie opened the pull request:
[contributing]: https://github.com/SparklingComet/TradeShop/blob/master/.github/CONTRIBUTING.md

## Pull Request Etiquette

There are several guidelines you should follow in order for your Pull Request to be merged.
Replace `[ ]` with `[x]` to cross the checkbox.

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

> It is sometimes better to include more changes in a single commit. 
  If you find yourself having an overwhelming amount of commits, you
  can **rebase** your branch.

## Changes

_Replace this sentence with a description of che changes introduced by this PR._

KillerOfPie committed:
Re-org classes by usage

KillerOfPie committed:
Re-org classes by usage + Missed package-info

KillerOfPie committed:
Separation of CommandRunner.java and DRY'd Shop item methods

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/framework/events/HopperShopAccessEvent.java:
```diff
@@ -22,7 +47,7 @@ public enum HopperDirection {
 	private Shop shop;
 	private boolean isForbidden;
-	private HopperDirection dir;
+	private final HopperDirection dir;
```
I didn't do this on purpose, if it needs to be changed back let me know

KillerOfPie replied:
I resolved all the other comments that were duplicates of this.

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/framework/events/PlayerPreTradeEvent.java:
```diff
@@ -45,10 +45,10 @@
-    private List<ShopItemStack> product;
-    private List<ShopItemStack> cost;
-	private Block clickedBlock;
+	private final Shop shop;
```
Same here

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/framework/events/PlayerPrepareTradeEvent.java:
```diff
@@ -45,12 +45,12 @@
-    private List<ShopItemStack> product;
-    private List<ShopItemStack> cost;
-	private Block clickedBlock;
+	private final Shop shop;
```
and here

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/framework/events/PlayerShopChangeEvent.java:
```diff
@@ -1,21 +1,46 @@
+
 	private static final HandlerList handlers = new HandlerList();
-	private Shop shop;
+	private final Shop shop;
```
^

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/framework/events/PlayerShopCreateEvent.java:
```diff
@@ -14,7 +39,7 @@ public class PlayerShopCreateEvent extends PlayerEvent implements Cancellable {
 	private static final HandlerList handlers = new HandlerList();
 	private boolean cancelled;
-	private Shop shop;
+	private final Shop shop;
```
^

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/framework/events/PlayerShopDestroyEvent.java:
```diff
@@ -39,7 +39,7 @@ public class PlayerShopDestroyEvent extends PlayerEvent implements Cancellable {
 	private static final HandlerList handlers = new HandlerList();
     private boolean cancelled, destroyBlock;
-	private Shop shop;
+	private final Shop shop;
```
^

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/framework/events/PlayerShopInventoryOpenEvent.java:
```diff
@@ -8,16 +33,16 @@
+
 	private static final HandlerList handlers = new HandlerList();
-	private Shop shop;
+	private final Shop shop;
```
^

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/framework/events/PlayerShopOpenEvent.java:
```diff
@@ -1,19 +1,44 @@
+
 	private static final HandlerList handlers = new HandlerList();
-	private Shop shop;
+	private final Shop shop;
```
^

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/shop/ShopChest.java:
```diff
@@ -50,7 +52,7 @@
 public class ShopChest extends Utils {
 
-	private final transient static TradeShop plugin = (TradeShop) Bukkit.getPluginManager().getPlugin("TradeShop");
+	private final static TradeShop plugin = (TradeShop) Bukkit.getPluginManager().getPlugin("TradeShop");
```
Transient not needed for static fields apparently and the newer version of Idea catches this.

SparklingComet replied:
I was wondering what this is about

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/shop/ShopSign.java:
```diff
@@ -53,7 +54,8 @@ enum Signs {
     WARPED_WALL_SIGN("1.16.0", "");
 
-    private List<Integer> minVer = Arrays.asList(new Integer[3]), maxVer = Arrays.asList(new Integer[3]);
+    private final List<Integer> minVer = Arrays.asList(new Integer[3]);
```
Final thing, seems the version i'm on checks and auto formats this in. I would assume if it does so then they could be final so none of these should cause problems.

KillerOfPie commented on src/main/java/org/shanerx/tradeshop/shop/listeners/ShopCreateListener.java:
```diff
@@ -119,14 +119,14 @@ public void onSignChange(SignChangeEvent event) {
 
-		if (cost != null && shop.getCost().isEmpty())
-			shop.setCost(cost);
+		if (cost != null && !shop.hasSide(ShopItemSide.COST))
```
Should I replace these with an `isSideEmpty1 method so that they are more readable? the method would just return this logic so its completely unnecessary but would be easier to read/check.

KillerOfPie committed:
DRY'd Shop User methods

KillerOfPie committed:
DRY'd Shop Constructors

KillerOfPie committed:
Fix for #128

KillerOfPie committed:
Implementation of #129

KillerOfPie committed:
Removed depreciated UNSUCCESSFUL_SHOP_MEMBERS Message

KillerOfPie committed:
Fix for #130

KillerOfPie committed:
Feature add for #131

KillerOfPie committed:
Feature Improvement for #132
- Improvements to Setting Section comparison code(due to errors when ne ones were added) should be a lot beefier and handle any future additions without problems
- Added configurable default item settings and ability to disable editing of individual item settings
- Added type checking to ObjectHolder
- Moved State String generation to ShopItemStack
- ShopItemStack Settings are ignored if not set to UserEditable in the config(uses default setting instead)

KillerOfPie committed:
Bug Fix for #134
- Moved #createShop to Utils
- Made #createShop more generic so sign creation and commands could both use it
- Added SHOP_CREATION Debug Level at bit 11(1024)
- Added toDebug method to Shop to allow a easily readable output of the shop
- Re-wrote updateSign methods to make them simpler to understand and allow sign/event passing from #createShop
- Removed event storing and related methods from Shop as it was unnecessary
- Deprecated #b64OverstackFixer since it is only used in legacy code
- Fixed some spelling in comments(the green underlines were bothering me)
- Removed unused `failed` methods from Utils

KillerOfPie committed:
Feature Add and Bug Fixes
- Added configurable colours for the middle lines on trade sides per material type #135
- Removed Message call in CommandRunner that should never be used and would never actually send if it was.
- Added SET_MANAGER and SET_MEMBER commands to the autofill for player names
- Added Config support for map values so individual setting enums are not required for settings that may change(like sign types), This should automatically add new items to existing maps as well
- Changed SettingSection weights to start at 50 increments based on nest depth to make them more easily readable
- Changed Member and Manager Lists into Sets to prevent Duplicates(hopefully)
- Added Shop#saveShop with a boolean parameter so that conditional saves don't need a separate if statement
- Removed Wall sign types since each sign type is handled the same whether it is on a wall or not
- Adjusted sign code to properly handle above
- Fixed error in ShoptradeListener where shops in the OOS state but without a product would try to check on stock.

KillerOfPie committed:
Bug Fixes
- Added conditional saving for Setting and Message files
- Added thorough change detection for setting defaults and upgrades to prevent unnecessary saves and fix bug preventing changes to default sign colours to be changed from their default values
- Fixed bug preventing setUser commands from working due to failed removal(which wasn't necessary for the set to process)
- Added missing return in ShopTradeListener which caused `OOS` code to run after `Incomplete` code in a switch

KillerOfPie committed:
Permission improvements
- Additions for #138

KillerOfPie committed:
Bug Fix
- Fix for my stupid debug output

KillerOfPie committed:
Bug Fix and Improvements
- Fix for more stupid mistakes
- Removed `Internal Permissions` as it is replaced by `Registered Permissions` #138
- fixed phrasing in `NO_TRADE_PERMISSION` pre-comment and `NO_TS_CREATE_PERMISSION` default message
- Prevent NONE permission from registering
- Added Permission registration debug to STARTUP level
- Fixed many places where getUsersUUID was not provided with which roles to get users for
- Fixed not checking for Create permission in CreateShop method allowing anyone to create and shop type by placing the sign

KillerOfPie committed:
Bug Fix and Improvements
- Remove `Player Level` command

KillerOfPie committed:
Bug Fix
- Fix for double message when shop could not be found during ShopUserCommands
- Added all as an option for targeting all shops when using ShopUserCommands

KillerOfPie committed:
Mapped Setting Improvements
- Mapped Setting now supports nested maps(so section > setting groups can be generated) this has no limit but cannot be changed externally so it should be fine and not encounter any un-ending loops
- Changed Per-Item settings to use Mapped values instead of individual settings and sections
- Changed ShopItemStackSettingKeys to load settings on each get call rather than when enums are created(since I think the latter could have problems when reloading)

KillerOfPie committed:
Per Shop Setting and necessary bug fixes
- Remove Debugger from utils to prevent early access
- Add in server default and user-editable settings for per-shop settings including hopper import/export and no-cost
- Update old values into new locations including no-cost-text, no-cost-amount, and existing hopper settings
- Fix for improper mapped settings writing
- Remove hopper settings from Metrics, need to add back before 2.6.0
- Add in asType settings to ObjectHolder so conversion code doesn't need to be added after retrieval.
- Changed old makeReadable Code to utilize WordUtils.
- Removed getString from ShopType to avoid early setting calls, replace with Setting enum value
- Add try/catch to skippableHoppers boolean updater to avoid early settings call
- Call updateSkippableHoppers a second time after settings has been fully initiated
- Add ShopSettingKeys
- Add Edit Settings GUI to `/ts edit` menu
- Added skippableShop cache so shops can be handled with a different time values and be removed when  their settings are changed and should be reprocessed.
- Added try/catch for IlleaglWorldException to PlayerSetting#getInvolvedStatus methods to prevent error when trying to load shops from unloaded worlds
- Fix for null appearing on shop sign lines 2/3 due to color being added
- Change availableTradeAmount for iTrade Shops to be 999
- Change updateStatus to change to OPEN when availableTrade > 0
- Change getFullTradeCount to check if costs can be traded on BiTrade Shops when no Product is in stock
- Change Hopper protection to not skip shop hoppers(code is commented out dependent on input)
- Add delayed Shop sign and available trade update on inventory close(when it is a shop inventory)
- Changed trade and trade check code to support new noCost setting

KillerOfPie committed:
Bug Fix
- Fix for bug reported by LewisB1#9506 on Discord where errors were being caused by old shops that required `/ts what` to view.

KillerOfPie committed:
Bug Fixes
- Proper fix for null appearing on signs(was due to the new sign color settings not being loaded)
- Changed serialized value for ObjectHolder so data storage is more readable
- Fix for bad data on a shop sign(That side of the trade will more than likely be removed)
- Dry'd Line formatting on signs for lines 2/3
- Added serialized name for `itemSettings` to properly read in the object with the old `shopSettings` name
- Added method to add ItemStacks to an inventory while temporary setting the inventories Max Stack Size to prevent overstacking items(May add as a shop setting in the future)
- Added updateFullTradeCount and updateSign to what command

SparklingComet committed:
Fix midding static handler method

SparklingComet committed:
Add PlayerShopChangeEvent#getChange()

KillerOfPie committed:
Pom.xml and plugin.yml changes

KillerOfPie committed:
Merge remote-tracking branch 'origin/2.6.0' into 2.6.0

KillerOfPie committed:
Bug Fixes
- Fix for Members displaying in shops and Member & Manager.
- Fix for `/ts removeUser` not filling TabComplete for players.
- Fix for double space in Message output between the Prefix and Message.

KillerOfPie committed:
Bug Fixes & POM Updates
- Updated plugin versions in POM.xml
- Fix for non DATA_ERROR debug output

KillerOfPie committed:
Bug Fixes & POM Updates
- Moved inventoryGUI to a local directory instead of it's own to prevent potential conflicts
- Fix for add Error during shop creation when adding users.

KillerOfPie committed:
Additional Debug

KillerOfPie committed:
Bug Fix & Command Improvement
- RemoveSide commands now default to the only item if no index was specified and the shop only has one item on the side
- Fix for members showing up as managers in the         edit menu
- Fix for managers not showing in edit menu

SparklingComet committed:
trivial optimization/refactoring of JsonShopConfiguration

KillerOfPie committed:
Bug Fix

KillerOfPie committed:
Catching up non SQL code from SQL branch 1/?

KillerOfPie committed:
Minor Changes
- Add SQLITE option to data-storage-type pre-comment
- POM Version updates
- POM Improvements
- Added `Clean` and `skipDocs` profiles for maven builds
- Moved `original` and JavaDocs output to `target/output-Files` sub folders
- Updated JavaDoc exclusions
- Excluded `META-INF/maven` files from builds

KillerOfPie committed:
Removal of un-needed legacy data fixes

KillerOfPie committed:
Catching up non SQL code from SQL branch 1/?

KillerOfPie committed:
Small reformat of lang file

KillerOfPie committed:
Implement Relative Directions for Chest checking
- Minor Testing done seems Okay
- Adds support for checking all 26 spots around a sign
- Checks for Chests/Signs based on the direction for the block
- Directions in the config should be formatted as `dir1+dir1+dir3` i.e `BACK` | `BACK+DOWN` | `LEFT+DOWN+BACK`
- There is no limit to the amount of directions(other than not being able to repeat a direction) but only 3 are needed to check specif the surrounding 26 blocks

SparklingComet commented on src/main/java/org/shanerx/tradeshop/data/config/Setting.java:
```diff
@@ -69,7 +69,7 @@ public enum Setting {
 
     // Global Options
-    ALLOWED_DIRECTIONS(SettingSection.GLOBAL_OPTIONS, "allowed-directions", new String[]{"DOWN", "WEST", "SOUTH", "EAST", "NORTH", "UP"}),
+    CHEST_DIRECTIONS(SettingSection.GLOBAL_OPTIONS, "chest-directions", new String[]{"BACK", "DOWN", "LEFT", "RIGHT", "UP", "FRONT"}),
```
It would probably be good to have a comment explaining what this is

KillerOfPie replied:
I think adding comments into these enums would start getting messy but we can if you think it's necessary. The lang has comments for the config output already. This did remind me to update the allowed-directions in lang to chest-directions though...

SparklingComet replied:
We can leave it for now and add a comment if people are confused

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/ListManager.java:
```diff
@@ -228,22 +223,23 @@ private void setGameMatList() {
+			try {
+				directions.add(RelativeDirection.valueOf(str));
+			} catch (IllegalArgumentException | NullPointerException ignored) {
+			}
```
I would actually throw an exception as the admin most likely misconfigured stuff.
If throwing is too dramatic, at the very least log a fat warning to the console.

KillerOfPie replied:
Exception is definitely too dramatic, I'll add a log message for it though.

KillerOfPie replied:
Actually, no this exception is only triggered by blanks or null being read(which shouldn't really happen anyways). this is more likely to be caused by owners improperly formatting then entering a value wrong, so I think a silent fail is better.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/Utils.java:
```diff
@@ -266,29 +268,51 @@ public String colorize(String msg) {
+				// 	If the relative block wasn't the ShopSign and we are checking a FlatFace, Then check if it is the other half of this Chest
+			} else if (flatFaces.stream().anyMatch((face) -> relativeOffset.getXOffset() == face.getModX() ||
+					relativeOffset.getYOffset() == face.getModY() ||
+					relativeOffset.getZOffset() == face.getModZ()) && (chest.getType().equals(Material.CHEST) || chest.getType().equals(Material.TRAPPED_CHEST))) {
```
I'm confused by what's going on here. Why are these materials hard-coded? Other things are.also possible...

KillerOfPie replied:
This code is specifically for double chests which(as far as I know) can only be Chest and Trapped_Chest

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/relativedirection/LocationOffset.java:
```diff
@@ -0,0 +1,62 @@
+/*
+ *
+ *                         Copyright (c) 2016-2019
```
We need to update this :S

KillerOfPie replied:
Yeah I keep thinking about it and forgetting

SparklingComet commented on src/main/java/org/shanerx/tradeshop/utils/relativedirection/RelativeCardinals.java:
```diff
@@ -0,0 +1,75 @@
+
+        return temp;
+    }
+}
```
I think this should be nested within `RelativeDirection` as `RelativeDirection.Cardinal`

KillerOfPie replied:
Well I was going to say that I don't like nesting when its used by classes outside the nest, but it isn't.

SparklingComet replied:
I agree with you but I think it's in order for small helper classes that are only used with/by one class

KillerOfPie committed:
Requested Changes + missed lang update
- Moved RelativeCardinals to be nested in RelativeDirection
- Change allowed-directions in lang to chest-directions
- Updated comment for above to be more descriptive of usage and options

KillerOfPie committed:
Catching up non SQL code from SQL branch 1/?

KillerOfPie committed:
Catching up non SQL code from SQL branch 1/?

KillerOfPie committed:
Catching up non SQL code from SQL branch 1/?

KillerOfPie committed:
Catching up non SQL code from SQL branch 1/?

KillerOfPie committed:
Fix for iTrades showing as full

KillerOfPie committed:
Fixes for #145

KillerOfPie committed:
Fixes for #145

KillerOfPie committed:
Dependency updates

KillerOfPie committed:
Dependency updates

KillerOfPie committed:
Fix for iTrade shops claiming to be full during trades + Spelling

KillerOfPie committed:
Fix for #143

KillerOfPie committed:
Fix for #143

KillerOfPie committed:
Changing to unified Gson code
- Added ConfigurationSerializationAdapter.java to handle MC object that use ConfigurationSerialization(Like ItemStack and its sub objects)

KillerOfPie committed:
Another attempt to fix #143
Added shop to call for creating item screen and used to save the changes directly from its save button rather than hoping it gets passed but to be saved.

KillerOfPie committed:
Pom Update for additional Jenkins options

KillerOfPie committed:
Minor Fixes and Cleanup

KillerOfPie committed:
Maybe fix for #143

KillerOfPie committed:
Fixing minor mistakes that caused errors

KillerOfPie committed:
Proper fix for #143

KillerOfPie committed:
Fix for Missing Serialization Data Attempt #1

KillerOfPie committed:
Fix for Missing Serialization Data Attempt #2

KillerOfPie committed:
Fix for Missing Serialization Data Attempt #2-AddDebug

KillerOfPie committed:
Fix for Missing Serialization Data Attempt #3

KillerOfPie committed:
Fix for Premature Expiration

KillerOfPie committed:
Debug Expansion and R-Add B64 itemStack Storage
Expanded Serialization debug, added B64 back to serialization to keep back compat, will probably remove before release.

KillerOfPie committed:
Bug Fixes

- Fix for MetaData not loading properly due to integers being loaded as Doubles when in a map
- Fix for pre-mature expiration

KillerOfPie committed:
Merge pull request #148 from Tradeshop/2.6.0-MissingEnchantments

Bug Fixes

KillerOfPie committed:
Bug Fixes
- Fix for lag spikes caused by Metrics
- Fix for ShopItemStacks that don't have itemSettings for whatever reason.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/CommandRunner.java:
```diff
@@ -1,1148 +0,0 @@
-			return null;
-		}
-	}
-}
```
this is a good one

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/BasicTextCommand.java:
```diff
@@ -0,0 +1,103 @@
+    public void setup() {
+        Message.SETUP_HELP.sendMessage(pSender, new Tuple<>(Variable.HEADER.toString(), Setting.TRADESHOP_HEADER.getString()));
+    }
+}
```
I like this

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/CommandRunner.java:
```diff
@@ -0,0 +1,95 @@
+
+	//------------------------------------------------------------------------------------------------------------------
+	//endregion
+}
```
This class is kinda pointless now. We should move findShop to ShopUser#findShop() since findShop does not depend on the commandPass etc, and get rid of this file.

KillerOfPie replied:
This class can still be useful in centralizing command code, a couple things still happen outside of findshop. Like breaking out pSender which most of the command classes use. I don't think I fully centralized everything that could be so this will probably have more in later versions outside of just providing a super constructor.

This findShop method is only used by commands since it needs a player line of sight. It can be moved to be a static method in shop, but I think it fits in a little better here especially since it can be protected and kept only for command classes and shop is already massive.

Either way, no reason to change now we can discuss and change oater if we want to.

SparklingComet replied:
`breaking out pSender` should be done in `CommandCaller` and passed as a constructor argument. `findShop` should definitely go into `ShopUser` for it is an action taken upon the user. It might be useful to some addons as well, and having it there makes more sense I think.

KillerOfPie replied:
pSender Breakout was moved to CommandPass since that's where we pull all the other command variables from.

findshop has been moved to ShopUser and renamed to findObservedShop, since I will also be adding a location based one rather than sight(probably also in ShopUser),

Leaving CommandCaller in, I might user it for something later, but the biggest reason is just to group all the commandcaller objects together. Was going to do an interface, but then I wouldn't be able to have required and predefined variables and would have to do that in each class.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/CreateCommand.java:
```diff
@@ -0,0 +1,116 @@
+
+        createShop(sign, pSender, ShopType.ITRADE);
+    }
+
```
Remove these three methods and create one method
```java
public void createShop(ShopType st) {
     // ...
     createShop(sign, pSender, st);
}
```

KillerOfPie replied:
yeaaahhh, not sure what I was tripping on when I made this. For the releases sake I'll change this for 2.6.1 since it makes no functional/performance difference

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/CreateCommand.java:
```diff
@@ -0,0 +1,116 @@
+
+    //------------------------------------------------------------------------------------------------------------------
+    //endregion
+}
```
See comment about CommandRunner#findShop, samething applies here.

KillerOfPie replied:
See my comment there, although if we leave findShop I will move findSign there to match. 

To add onto that, moving this the shop would require us to either pass psender and still perform the null check or pass commandPass and do the player check, cast and null check

SparklingComet replied:
the pSender is a Player, not CommandSender, so no player check necessary. We should accept it as a constructor parameter, passed by CommandCaller. The thing can be expected to not be null since the checks are done centrally by CommanPass. Checks of kind

```
        if (pSender == null) {
            Message.PLAYER_ONLY_COMMAND.sendMessage(pSender);
            return null;
        }
```

should be done inside CommandCaller as well. It might be cleaner to have one function there that is used for every "player-only" command, as that is where we should be retrieving the pSender. That way we only have this code once instead of repeating it in each command class.

KillerOfPie replied:
Player check was already happening in command caller. I only found one in findSign, so I probably just missed that when I added the central player check.

```java
switch (command.checkPerm(sender)) {
			case NO_PERM:
				Message.NO_COMMAND_PERMISSION.sendMessage(sender);
				return true;
			case PLAYER_ONLY:
				Message.PLAYER_ONLY_COMMAND.sendMessage(sender);
				return true;
		}
```

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/GUICommand.java:
```diff
@@ -0,0 +1,293 @@
+            EDIT_LAYOUT = {"aggggggga", "ap c s na"},
+            ITEM_LAYOUT = {"u ggggggg", "j hhhhhhh", "ap cbs na"},
+            SETTING_LAYOUT = {"ggggggggg", "hhhhhhhhh", "ap cbs na"},
+            WHAT_MENU = {"141125333", "1qqq2eee3", "11p123n33"};
```
What in the world is this? ALso, if they are only used in EditCommand, they should be defined there.

KillerOfPie replied:
They are layout formats for the GUIs, pretty sure some are used here and in the what command as well. Either way want to keep them in one place so they don't get unnecessarily repeated.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/GUICommand.java:
```diff
@@ -0,0 +1,293 @@
+    protected List<ShopItemStack> costItems = new ArrayList<>(),
+            productItems = new ArrayList<>();
+    protected List<Boolean> costItemsRemoval = new ArrayList<>(),
+            productItemsRemoval = new ArrayList<>();
```
Do these lists need to be global? Aren't they one-use only?

KillerOfPie replied:
They do because the update isn't happening in the local area that the changes are(since each button has its own lambda method). so each button needs to be able to access ad make changes and the update button needs to be able to read.

With the GUI code a lot of it comes down to whether or if it works because debugging it after changes can be excessively difficult.

KillerOfPie replied:
Idea doesn't seem to care about it being moved but since the GUI stuff is so finicky, I want to save this for a release that will have more thorough testing. I also want time to see if I can write the GUI code cleaner.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/EditCommand.java:
```diff
@@ -0,0 +1,334 @@
+                                            "true",
+                                            getBooleanItem(true),
+                                            key.makeReadable(),
+                                            "State: True"
```
bad indentation

KillerOfPie replied:
See above, resolving

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/EditCommand.java:
```diff
@@ -0,0 +1,334 @@
+                                            "false",
+                                            getBooleanItem(false),
+                                            key.makeReadable(),
+                                            "State: False"
```
bad indentation

KillerOfPie replied:
See above, resolving

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/EditCommand.java:
```diff
@@ -0,0 +1,334 @@
+                    if (state.isBoolean()) {
+                        if (key.isUserEditable(shop.getShopType())) {
+                            changeGroup.addElement(new GuiStateElement('e',
+                                    state.asBoolean().toString(),
```
8 spaces? should be 2 tabs

KillerOfPie replied:
~~I think I did this on purpose to distinguish between the if indentation and a continued line(or idea did it on its own) this code is disgusting to read and deal with so tat minor change can make a huge difference in readability. I think I stayed pretty consistent on this formatting style and my tabs are set to spaces since they are more consistent.~~

~~Will look at in 2.6.1 code and see how big the difference is/if idea did that on its own.~~

~~Some of the below ones are excessive and will be fixed.~~

This formatting is how Idea wants it above reason still apply but it'll stay because every time it gets auto formatted it will go back.  Also not going to make a commit for one extra empty line. Will remove on my next commit if I remembe.

SparklingComet replied:
discard my comment here, 2 layers makes 100% sense for continuing the previous statement. I think there are some places however, where there is more than 1 layer in if-body sort of situations.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/EditCommand.java:
```diff
@@ -0,0 +1,334 @@
+                            changeGroup.addElement(new StaticGuiElement('e', getBooleanItem(state.asBoolean()), key.makeReadable(), value.toString()));
+                        }
+
+
```
too much whitespace, please remove

KillerOfPie replied:
See above, resolving

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/EditCommand.java:
```diff
@@ -0,0 +1,334 @@
+                                    user.getName(),
+                                    "Position: NONE",
+                                    "Click to cycle the player to Manager.")
+                    ));
```
is there no less repetitive way of writing this?

KillerOfPie replied:
not really that I know of, I could probably combine 
``` java
shopUsers.remove(user);
user.setRole(ShopRole.MEMBER);
 shopUsers.add(user);
```
into a method but that's it, the rest are parameters tat need to be passed. This is part of the reason this code is annoying to deal with.

SparklingComet replied:
hmm fair enough, not really worth making a function I guess

KillerOfPie replied:
Will attempt when I rewrite GUI code later

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/GUICommand.java:
```diff
@@ -0,0 +1,293 @@
+                            item.getStateString(new ObjectHolder<>(1))
+
+                    ));
+        }
```
bad indentation; also is there no way of writing this in a more concise  and clear way?

KillerOfPie replied:
see earlier indentation comment.

Can't really simplify this since one index number needs to match the value we save(which needs to be an int otherwise dealing with it in the config requires more effort and it takes more space to store) and the other needs to match array values. for repetition see previous comment.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/GUICommand.java:
```diff
@@ -0,0 +1,293 @@
+        }
+
+        return new StaticGuiElement('e', getBooleanItem(item.getShopSetting(setting).asBoolean()), setting.makeReadable(), item.getStateString(setting));
+    }
```
bad indentation in this function

KillerOfPie replied:
See previous, resolving

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/ShopCommand.java:
```diff
@@ -0,0 +1,145 @@
+                    Message.NO_COMMAND_PERMISSION.sendMessage(pSender);
+                    return;
+                }
+        }
```
Switch is missing break;, accidental fallthrough if player has CreteBi permission

Remove switch, one if is enough:

```java
if (!Permissions.hasPermission(pSender, shop.getShopType() == ShopType.TRADE ? Permissions.CREATE : Permissions.CREATEBI)) {
    Message.NO_COMMAND_PERMISSION.sendMessage(pSender);
    return;
}
```

Also should we add a check to make sure shopType is not ITRADE or whatever?

KillerOfPie replied:
This doesn't get called if shop is iTrade, can combine in 2.6.1 as the fall through here would actually matter too much. also it checks the opposite permission of the current type

KillerOfPie replied:
added breaks

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/ShopCommand.java:
```diff
@@ -0,0 +1,145 @@
+        }
+
+        if (!(shop.getUsersUUID(ShopRole.MANAGER, ShopRole.OWNER).contains(pSender.getUniqueId())
+                || (Setting.UNLIMITED_ADMIN.getBoolean() && Permissions.isAdminEnabled(pSender)))) {
```
I think it would be clearer if the condition was refactored slightly: !(a || b) <=> !a && !b

KillerOfPie replied:
this is !a || (b && c), does need to change to !(a || (b && c)) although this doesnt matter now since admin cant switch shop types

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/GeneralPlayerCommand.java:
```diff
@@ -0,0 +1,109 @@
+        } else {
+            int amount = Setting.MULTI_TRADE_DEFAULT.getInt();
+
+            if (isInt(command.getArgAt(1)))
```
What is isInt?

KillerOfPie replied:
checks if a string is an integer, although this functionality moved to object holder so this should be redone in 2.6.1

SparklingComet replied:
I figure, but I am confused as to where this function is defined

KillerOfPie replied:
Currently in Utils, again needs to move to using the ObjectHolder implementation that I added

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/GeneralPlayerCommand.java:
```diff
@@ -0,0 +1,109 @@
+            if (amount < 2)
+                amount = 2;
+            else if (amount > Setting.MULTI_TRADE_MAX.getInt())
+                amount = Setting.MULTI_TRADE_MAX.getInt();
```
amount = Math.min(Math.max(2, amount), Setting.MULTI_TRADE_MAX.getInt());

KillerOfPie replied:
ummm yea, i have that somewhere else, i wonder why not here...

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/ShopCommand.java:
```diff
@@ -0,0 +1,145 @@
+            return;
+
+        if (!(shop.getUsersUUID(ShopRole.MANAGER, ShopRole.OWNER).contains(pSender.getUniqueId())
+                || Permissions.isAdminEnabled(pSender))) {
```
also check if admin mode is enabled generally, like you did on L136

KillerOfPie replied:
This does check admin mode and shouldnt be affected by unlimited admin

KillerOfPie replied:
This is good, unlimited admin is for thing we wouldn't have an admin check since I didn't find it to be something they should need.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/ShopCommand.java:
```diff
@@ -0,0 +1,145 @@
+            return;
+
+        if (!(shop.getUsersUUID(ShopRole.MANAGER, ShopRole.OWNER).contains(pSender.getUniqueId())
+                || Permissions.isAdminEnabled(pSender))) {
```
same here (?)

KillerOfPie replied:
```suggestion
                || Permissions.isAdminEnabled(pSender))) {same as above
```

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/ShopCommand.java:
```diff
@@ -0,0 +1,145 @@
+            return;
+
+        if (!Permissions.hasPermission(pSender, Permissions.EDIT)) {
+            Message.NO_COMMAND_PERMISSION.sendMessage(pSender);
```
Admin mode should be supported here

KillerOfPie replied:
I saw no reason for an admin to change a shop between trade an bitrade, but we can add the unlimted and admin enabled check if you'd like

SparklingComet replied:
I think it would be fine as part of the unlimited admin.

KillerOfPie replied:
added for unlimited admin

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/ShopUserCommand.java:
```diff
@@ -0,0 +1,260 @@
+    public String toString() {
+        return text.toString();
+    }
+}
```
is this a github bug or something?why is this enum at the top level class declaration of ShopUserCommand.java

KillerOfPie replied:
Not sure, I more tan likely wouldn't have put it there, will fix in 2.6.1 if it is actually this way. No functional or performance difference

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/ShopUserCommand.java:
```diff
@@ -0,0 +1,260 @@
+                else
+                    managers.append(", ").append(usr.getName());
+            }
+        }
```
Replace this for loop with String.join() please

KillerOfPie replied:
String.join would still need a loop since the array is shopUser objects not strings(This would actually require 2 loops vs the one here since String.join needs to iterate across the collection that was already converted to strings) . Not sure why im assigning new StringBuilders instead of just appending the first item. Also don't think String.join is ay more efficient than StringBuilder

I think this can be saved for 2.6.1 though. Up to you though since I think it'll most like be a minor cleanup/reformat(can probably use a Collections.forEach or other lambda loop to accomplish this.

SparklingComet replied:
> Up to you though since I think it'll most like be a minor cleanup/reformat(can probably use a Collections.forEach or other lambda loop to accomplish this.

Yes this would be better I think. There would not be any performance benefit, it is just easier to read that way I  think

KillerOfPie replied:
I'm stupid, here's the rewrite. apparently i forgot I had a getUserNames method or I didn't updaate this after making that -_-

```java
public void who() {
        Shop shop = ShopUser.findObservedShop(command.getPlayerSender());

        if (shop == null)
            return;

        if (shop.getShopType().isITrade()) {
            Message.WHO_MESSAGE.sendMessage(command.getPlayerSender(),
                    new Tuple<>(Variable.OWNER.toString(), Setting.ITRADESHOP_OWNER.getString()),
                    new Tuple<>(Variable.MANAGERS.toString(), "None"),
                    new Tuple<>(Variable.MEMBERS.toString(), "None"));
            return;
        }

        String owner = String.join(", ", shop.getUserNames(ShopRole.OWNER));
        String managers = String.join(", ", shop.getUserNames(ShopRole.MANAGER));
        String members = String.join(", ", shop.getUserNames(ShopRole.MEMBER));
        
        Message.WHO_MESSAGE.sendMessage(command.getPlayerSender(),
                new Tuple<>(Variable.OWNER.toString(), owner.length() < 3 ? "None" : owner),
                new Tuple<>(Variable.MANAGERS.toString(), managers.length() < 3 ? "None" : managers),
                new Tuple<>(Variable.MEMBERS.toString(), members.length() < 3 ? "None" : members));
    }
```

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/ShopUserCommand.java:
```diff
@@ -0,0 +1,260 @@
+                else
+                    members.append(", ").append(usr.getName());
+            }
+        }
```
Same here

KillerOfPie replied:
see above, resolving

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/ShopUserCommand.java:
```diff
@@ -0,0 +1,260 @@
+            }
+        }
+
+        if (managers.toString().equals("")) {
```
no remove this and, add an else case `if (shop.hasUsers(ShopRole.MANAGER)) {`

KillerOfPie replied:
See above, these can totally be elses and creating a new stringbuilder isnt necessary. but would rather rewrite the whole section cleaner in 2.6.1

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/ShopUserCommand.java:
```diff
@@ -0,0 +1,260 @@
+        }
+        if (members.toString().equals("")) {
+            members = new StringBuilder("None");
+        }
```
same applies here

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/ShopUserCommand.java:
```diff
@@ -0,0 +1,260 @@
+        }
+
+        return null;
+    }
```
`target` should not be a global variable. `shopUserCommandStart` should accept `OfflinePlayer target` as parameter and not modify this value upon failure. As a matter of fact, target should not even be `OfflinePlayer`, but just `Player` (if they have played before but are offline, this thing blows up). Check if they are online, if so use `Bukkit.getPlayerExact(name)`.

KillerOfPie replied:
Offline player is much safer than player, since A) this needs to work even if the player is offline, and B) Offline player is more friendly for offline servers(If I remember correctly). Think you got it backwards since this would blow up if using player and the player is offline.

As for target, even with it global it is effectively a local argument since this is a new instance for each command that is run through it. Not really any benefit/difference to passing target rather than having it global.

SparklingComet replied:
yes I think you are right, was tripping there. Still, this should not indicate failure by setting global variables to null. Also I still think it is safer to make target local to prevent accidental modification in the future, and move to exception based failure.

KillerOfPie replied:
target gets passed to the method now.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/commands/commandrunners/ShopUserCommand.java:
```diff
@@ -0,0 +1,260 @@
+     * Checks if targeted player exists and if player is looking at a shop while not targetting all owned shops
+     *
+     * @param applyAllOwned Is Player targetting all owned shops
+     * @return shop if found or null if not needed; returning null while setting target to null indicates failure, command should respond with an immediate blank return.
```
This is not C lol, just  throw exceptions in case of failure. Don't set `target = null` to signify failure.

KillerOfPie replied:
not sure if I was doing something with the null on the other side, but I think dealing with null in this case is much more efficient then an exception. will look at for 2.6.1

SparklingComet replied:
I would think it would be more "efficient" on a sub-ms scale since we only unwind the stack by one call

KillerOfPie replied:
that took a bit to figure out what was happening. throws UnsupportedOperationException on failure. catches but ignores

SparklingComet commented on src/main/java/org/shanerx/tradeshop/data/config/ConfigManager.java:
```diff
@@ -92,40 +97,67 @@ public void load() {
                 break;
         }
+
+        return hasUpgraded.contains(true);
```
I don't understand this return value

KillerOfPie replied:
if a value is added byany of the addKeyValues then there will be a true in the set triggering the save

SparklingComet replied:
makes sense

SparklingComet commented on src/main/java/org/shanerx/tradeshop/data/config/ConfigManager.java:
```diff
@@ -92,40 +97,67 @@ public void load() {
         config = YamlConfiguration.loadConfiguration(file);
 
-        setDefaults();
+        return setDefaults();
```
What does this return value mean?

KillerOfPie replied:
passes return from 151 to 105 to trigger save

SparklingComet commented on src/main/java/org/shanerx/tradeshop/data/config/ConfigManager.java:
```diff
@@ -92,40 +97,67 @@ public void load() {
+     * @param shouldSave true if save should proceed
+     */
+    public void save(boolean shouldSave) {
+        if (shouldSave) save();
```
why is this necessary?

KillerOfPie replied:
its not, I just prefer it over the ifs everywhere I want a conditional save

SparklingComet commented on src/main/java/org/shanerx/tradeshop/data/config/ConfigManager.java:
```diff
@@ -92,40 +97,67 @@ public void load() {
         }
 
-        save();
+        save(hasUpgraded.contains(true));
```
I think this does not make much sense, better to keep it simple `if (hasUpgraded.contains(true)) save());`

KillerOfPie replied:
I think this is simpler then adding an if state to everywhere this happens

KillerOfPie replied:
This save was only called once, so I made this change.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/data/config/ConfigManager.java:
```diff
@@ -153,14 +185,14 @@ public void save() {
-                                outputMap.getOrDefault(section.getSection().getPath(), "") + section.getFileString()));
+                        Arrays.stream(SettingSection.values()).sorted(new CompareSettingSections()).forEach((section) -> outputMap.put(section.getPath().toLowerCase().replace("_", "-"), section.getFileString()));
+                        Arrays.stream(Setting.values()).forEach((setting) -> outputMap.put(setting.getSection().getPath().toLowerCase().replace("_", "-"),
+                                outputMap.getOrDefault(setting.getSection().getPath().toLowerCase().replace("_", "-"), "") + setting.getFileString()));
```
why is #getPath no longer being used?

KillerOfPie replied:
it is?

SparklingComet replied:
never mind that, I skipped over it.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/data/config/ConfigManager.java:
```diff
@@ -153,14 +185,14 @@ public void save() {
-                                outputMap.getOrDefault(section.getSection().getPath(), "") + section.getFileString()));
+                        Arrays.stream(MessageSection.values()).sorted(new CompareMessageSections()).forEach((section) -> outputMap.put(section.getPath().toLowerCase().replace("_", "-"), section.getFileString()));
+                        Arrays.stream(Message.values()).forEach((message) -> outputMap.put(message.getSection().getPath().toLowerCase().replace("_", "-"),
+                                outputMap.getOrDefault(message.getSection().getPath().toLowerCase().replace("_", "-"), "") + message.getFileString()));
```
same here

KillerOfPie replied:
^

SparklingComet commented on src/main/java/org/shanerx/tradeshop/data/config/Message.java:
```diff
@@ -315,6 +330,52 @@ public final void sendItemMultiLineMessage(Player player, Map<Variable, List<Ite
+                entry.getValue().forEach((k, v) -> {
+                    itemList.append("\n")
+                            .append(format.replace(Variable.SHOP.toString(), k).replace(Variable.STATUS.toString(), v));
+                });
```
String.join("\n", things)

KillerOfPie replied:
I don't think using String.Join has any benefit over multiple appends if We're still sing the stringbuilder.

SparklingComet replied:
I think complexity is the same (compare [StringBuilder](https://hg.openjdk.org/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/lang/AbstractStringBuilder.java) to [String.join](https://hg.openjdk.org/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/lang/AbstractStringBuilder.java)), this is more a matter of visibility, not that important

KillerOfPie replied:
After looking at this, String.Join could only be used if we didn't need to use the for loop. I don't see a way of doing that right now and would rather not fuss with fairly clean working code.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/data/config/Setting.java:
```diff
@@ -126,8 +138,9 @@ public static Setting findSetting(String search) {
-    static void upgrade() {
+    static boolean upgrade() {
         double version = CONFIG_VERSION.getDouble();
+        Set<Boolean> hasUpgraded = new HashSet<>(); // Uses this instead of a boolean to later replace below ifs with boolean return methods...
```
uhh... what?

KillerOfPie replied:
The alternative was repeatedly setting a Boolean, I think I changed because either Idea hated the fact that I just kept setting the boolean to true or it was annoying to read. Probably the second caused by the first. either way there will only ever be 1 true in the set sooo its not terrible and at least idea doesn't throw a hissy fit about it.

OOOOH strike that I think I planned on having upgrade call methods and returning a boolen to the set. If each method returned to a boolean it wouldn't track if anything was changed. In this case it doesn't matter what each returns, if there is a true in the set at the end there was a cange.

KillerOfPie replied:
The below ifs are stupid repetitive

SparklingComet commented on src/main/java/org/shanerx/tradeshop/data/config/Setting.java:
```diff
@@ -223,14 +306,40 @@ public String getPath() {
+                processed.append(localLead).append(key.toString()).append(":\n");
+                processed.append(processMapValue(((Map<?, ?>) value), localLead + leadIncrease));
+            } else {
+                processed.append(localLead).append(key.toString()).append(": ").append(new Yaml().dump(value));
```
please rewrite using String.format()

KillerOfPie replied:
not sure how you expect format to work here or be any cleaner, this needs to be able to turn nesting maps into the string output, and again the stringbuilder is still there so not sure about improvements

SparklingComet commented on src/main/java/org/shanerx/tradeshop/data/config/Setting.java:
```diff
@@ -223,14 +306,40 @@ public String getPath() {
+            keyOutput.append(section.getSectionLead()).append(getKey()).append(":\n");
+            keyOutput.append(processMapValue(((Map<?, ?>) defaultValue), section.getSectionLead() + leadIncrease));
+        } else {
+            keyOutput.append(section.getSectionLead()).append(getKey()).append(": ").append(new Yaml().dump(getSetting()));
```
same here, to hell with performance enhancements from StringBuilder, I think String.format makes it clearer

KillerOfPie replied:
lol you shouldve said that sooner before i said it 20 times. I prefer reading it this way than format....

SparklingComet commented on src/main/java/org/shanerx/tradeshop/item/ShopItemStack.java:
```diff
@@ -510,16 +507,48 @@ public String getCleanItemName() {
+            return "State: " + ret;
+        } catch (ClassCastException ex) {
+            return "State: ERROR";
+        }
```
what is this?

KillerOfPie replied:
this is text that shows on the gui items for shop itemstacksettings

SparklingComet commented on src/main/java/org/shanerx/tradeshop/item/ShopItemStack.java:
```diff
@@ -510,16 +507,48 @@ public String getCleanItemName() {
+                ret = "Unknown";
+
+            return "State: " + ret;
+        } catch (ClassCastException ex) {
```
does this actually happen though?

KillerOfPie replied:
I dont think so but I think idea thinks it can

SparklingComet replied:
fair enough

SparklingComet commented on src/main/java/org/shanerx/tradeshop/item/ShopItemStack.java:
```diff
@@ -510,16 +507,48 @@ public String getCleanItemName() {
+
+    public String getStateString(ObjectHolder<?> stateSetting) {
+        try {
+            String ret = "";
```
`String ret = null;`

KillerOfPie replied:
2.6.1

KillerOfPie replied:
Rewrote this whole method down to 18 lines w/ no switches

SparklingComet commented on src/main/java/org/shanerx/tradeshop/item/ShopItemStack.java:
```diff
@@ -510,16 +507,48 @@ public String getCleanItemName() {
+                }
+            }
+
+            if (ret.length() < 2)
```
Do the above,then `if (ref == null)`

KillerOfPie replied:
^

KillerOfPie replied:
Just changed initial value to Unknown so it's already set to that if nothing changes it.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/shop/Shop.java:
```diff
@@ -0,0 +1,1211 @@
+		if (!cost.isEmpty())
+			cost.forEach(costItem -> sb.append("       ").append(costItem.toConsoleText()).append("\n"));
+
+		return utils.colorize(sb.toString());
```
Please rewrite with String.format(), this StringBuilder#append stuff is kinda hard to read...

KillerOfPie replied:
do I have to though? also efficiency is important in debug code so the ugly is probably worth it

KillerOfPie replied:
Also going to leave this since adding all the calls to one action may make it harder to track down what caused an error.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/shop/Shop.java:
```diff
@@ -0,0 +1,1211 @@
+	 *
+	 * @param shouldSave true if save should proceed
+	 */
+	public void saveShop(boolean shouldSave) {
```
this should be removed; just use `if (shouldSave) saveShop();`

KillerOfPie replied:
I think in the cases this is used adding an if was either difficult or ugly or just happened too many times

KillerOfPie replied:
Looks like I was thinking of the config save rather than this one. 

I do still prefer this where it is used and will leave it in.

SparklingComet commented on src/main/java/org/shanerx/tradeshop/shop/Shop.java:
```diff
@@ -0,0 +1,1211 @@
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+public class Shop implements Serializable {
```
fix indentation depth

KillerOfPie replied:
2.6.1

SparklingComet commented on src/main/java/org/shanerx/tradeshop/shop/Shop.java:
```diff
@@ -0,0 +1,1211 @@
+					if (!excludedPlayers.contains(owner.getUUID()))
+						users.add(owner);
+					break;
+			}
```
too much repetition, would be cool if this could be reduced ( we have the same thing 3 times)

KillerOfPie replied:
could probably use lambda ifs but not sure what else...2.6.1

KillerOfPie replied:
I lied apparently, brought the whole method down to 8 lines w/ no repetition.
